Geometry_engine-issues-880,896-SplitAtPoints-SortAlongCurve-Distance-etc

LMarkowski opened the pull request:
Closes #880 
Closes #896 


### Test files
Test files are [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FBHoM%5FEngine%2FGeometry%5FEngine%2FGeometry%5Fengine%2Dissues%2D880%2C896%2DSplitAtPoints%2DSortAlongCurve%2DDistance%2Detc)


### Changelog
New methods added:
- SortAlongCurve: Arc, Circle, PolyCurve
- SplitAtPoints: Arc, Circle PolyCurve
- Normal: Arc

Rewrited code for new inputs:
- IsCoplanar: PoluCurve (rewrited from polylines)
- IsOnCurve: everything but nurbs and ellipse. (Refer to Distance())
- Join: List PolyCurve (rewrited from Join(List ICurve))
- CurveIntersections: PolyCurve, PolyCurve (rewrited from CurveIntersections(ICurve, ICurve))

Fixed:
- ClosestPoint(Arc,Point) now checks if the point is arc.centre and then always returns StartPoint. It caused bug in Distance() -> #896  . Previous method is deprecated and functionality is preserved.

### Additional comments
These are all methods which were necessery to create booleanMethods for PolyCurves (#874 ). They are coming in another PR right after this one get merged.

pawelbaran commented:
@al-fisher what's the current strategy with the test scripts? is it worth putting time and effort into getting the scripts templated, cleaned up and placed in the right test files?

LMarkowski commented:
@pawelbaran 
- added missing Tolerance references
- adjusted gh method and verify cluster, now it's ok. Wider explanation is in the .gh file
- fixed
- it was caused by BHoM's RandomPolyCurve. Fixed and changed shattering method for a better one.
